docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+Heapster Authentication and Authorization
+=========================================
+
+Current State
+-------------
+
+Heapster authentication is based on checking client certificates against a CA.
+Authorization is either just a static list of names, or non-existant.  In
+either case, anyone authenticated is authorized to perform any operation.  With
+push metrics, the proposed plan was to allow adding a second CA, to allow
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
```

Comment:
Somewhat.  The motivation comes from several related issues:

1. We want to be able to restrict who can push metrics, read recent metrics (via the model), and read historical metrics (via Oldtimer) -- it's desirable to allow different sets of people to have permissions for those things.
2. We *potentially* want to be able to restrict who can push metrics associated with certain "objects" (e.g. you might want some to be able to push metrics for any cluster objects, while other might be only allowed to push metrics in their own namespace).
3. We seen concern over the fact that you cannot restrict reading of metrics to a certain scope for different users.  A mechanism to restrict users to reading things only in their own namespace, for example, would allow Heapster access to be more "safely" granted to different applications and users on the cluster.